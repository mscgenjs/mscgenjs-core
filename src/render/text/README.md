# Rendering text: other script languages

To be able to switch between the languages mscgen_js supports, the code contains
functions to convert abstract syntax trees back to text and some snazzy
modules to manipulate syntax trees.

## Language rendering

The modules that (re-)render language code from abstract syntax trees
have a _lot_ in common. So much, that most individual language can be
expressed as a configuration on one processing module. We've taken the
most expansive of the languages (XÃ¹) as the base class and implemented
the other languages (and doxygen) as children.

- [ast2xu](ast2xu)
  - [ast2mscgen](ast2mscgen)
    - [ast2doxygen](ast2doxygen)
  - [ast2msgenny](ast2msgenny)

## Animation

:page_with_curl: code in [ast2animate.js](ast2animate.js)

## Language rendering: dot

Each sequence chart can be expressed as a communications diagram.
GraphViz dot is a textual language that can be used to specify
communications diagrams. ast2dot transforms an abstract syntax
tree into a dot program.

:page_with_curl: code in [ast2dot.js](ast2dot.js)
