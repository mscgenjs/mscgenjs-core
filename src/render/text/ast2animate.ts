import _cloneDeep from "lodash.clonedeep";
import * as mscgenjsast from "../../parse/mscgenjsast";

const EMPTY_ARC = [{kind: "|||" as mscgenjsast.ArcKindType}];
const EMPTY_AST = {
    entities: [],
    meta: {
        extendedArcTypes: false,
        extendedFeatures: false,
        extendedOptions: false,
    },
};

class FrameFactory {
    public AST: mscgenjsast.ISequenceChart;
    public arcs: mscgenjsast.IArc[][];
    public len: number;
    public noRows: number;
    public position: number;
    public frames: mscgenjsast.ISequenceChart[];
    public preCalculate: boolean;

    constructor(pAST: mscgenjsast.ISequenceChart, pPreCalculate?: boolean) {
        this.AST          = EMPTY_AST;
        this.arcs         = [[]] as mscgenjsast.IArc[][];
        this.len          = 0;
        this.noRows       = 0;
        this.position     = 0;
        this.frames       = [];
        this.preCalculate = false;

        if (pAST) {
            if (pAST && typeof pPreCalculate !== "undefined") {
                this.init(pAST, pPreCalculate);
            } else {
                this.init(pAST);
            }
        }
    }

    /*
        * initializes the frame generator with an AST and
        * calculates the number of frames in it.
        *
        * @param pAST - abstract syntax tree to calculate
        * @param pPreCalculate - if true the module will pre-calculate all frames
        *                 in advance. In all other cases the module will
        *                 calculate each frame when it is called for (with
        *                 getFrame/ getCurrentFrame calls). Note that the
        *                 latter usually is fast enough(tm) even for real
        *                 time rendering. It will probably save you some
        *                 cpu cycles when you're going traverse the frames
        *                 a lot, at the expense of memory usage.
        *
        *                 Paramater might get removed somewhere in the near
        *                 future.
        */
    public init(pAST: mscgenjsast.ISequenceChart, pPreCalculate?: boolean): void {
        this.preCalculate = pPreCalculate ? true === pPreCalculate : false;
        this.AST          = _cloneDeep(pAST);
        this.len          = this._calculateLength(pAST);
        this.noRows       = this._calcNumberOfRows(pAST);
        this.position     = 0;
        if (this.AST.arcs) {
            this.arcs     = _cloneDeep(this.AST.arcs);
            this.AST.arcs = [];
        }
        this.frames = [];
        if (this.preCalculate) {
            for (let i = 0; i < this.len; i++) {
                this.frames.push(_cloneDeep(this._calculateFrame(i)));
            }
        }
    }

    /*
     * Set position to the provided frame number
     * If pFrameNumber > last frame, sets position to last frame
     * If pFrameNumber < first frame, sets position to first frame
     */
    public setPosition(pPosition: number): void {
        this.position = Math.min(this.len, Math.max(0, pPosition));
    }

    /*
     * Go to the first frame
     */
    public home(): void {
        this.position = 0;
    }

    /*
     * Skips pFrames ahead. When pFrames not provided, skips 1 ahead
     *
     * won't go beyond the last frame
     */
    public inc(pFrames: number = 1): void {
        this.setPosition(this.position + pFrames);
    }

    /*
     * Skips pFrames back. When pFrames not provided, skips 1 back
     *
     * won't go before the first frame
     */
    public dec(pFrames: number = 1): void {
        this.setPosition(this.position - pFrames);
    }

    /*
     * Go to the last frame
     */
    public end(): void {
        this.position = this.len;
    }

    /*
     * returns the current frame
     */
    public getCurrentFrame(): mscgenjsast.ISequenceChart {
        return this.getFrame(this.position);
    }

    /*
     * returns frame pFrameNo
     * if pFrameNo >= getLength() - returns the last frame (=== original AST)
     * if pFrameNo <= 0 - returns the first frame (=== original AST - arcs)
     */
    public getFrame(pFrameNo: number): mscgenjsast.ISequenceChart {
        pFrameNo = Math.max(0, Math.min(pFrameNo, this.len - 1));
        if (this.preCalculate) {
            return this.frames[pFrameNo];
        } else {
            return this._calculateFrame(pFrameNo);
        }
    }

    /*
     * returns the position of the current frame (number)
     */
    public getPosition(): number {
        return this.position;
    }

    /*
     * returns the number of "frames" in this AST
     */
    public getLength(): number {
        return this.len;
    }

    /*
     * returns the ratio position/ length in percents.
     * 0 <= result <= 100, even when position actually exceeds
     * length or is below 0
     */
    public getPercentage(): number {
        return (this.len > 0) && (this.position > 0) ? 100 * (Math.min(1, this.position / this.len)) : 0;
    }

    /*
     * returns the number of rows for the current AST
     */
    public getNoRows() {
        return this.noRows;
    }

    private _drawArcsUntilRow(pFrameNo: number) {
        let lFrameCount = 0;
        let lRowNo = 0;
        if (this.AST.arcs) {
            while (lFrameCount < pFrameNo) {
                this.AST.arcs[lRowNo] = [] as mscgenjsast.IArc[];
                for (let j = 0; (j < this.arcs[lRowNo].length) && (lFrameCount++ < pFrameNo); j++) {
                    this.AST.arcs[lRowNo].push(this.arcs[lRowNo][j]);
                }
                lRowNo++;
            }
        }
        return lRowNo;
    }

    private _fillDownWithEmptyArcs(pRowNoFrom: number) {
        if (this.AST.arcs) {
            for (let k = pRowNoFrom; k < this.noRows; k++) {
                this.AST.arcs[k] = EMPTY_ARC;
            }
        }
    }

    /*
     * Returns the AST the subset frame pFrameNo should constitute
     */
    private _calculateFrame(pFrameNo: number): mscgenjsast.ISequenceChart {
        pFrameNo = Math.min(pFrameNo, this.len - 1);

        if (this.len - 1 > 0) {
            this.AST.arcs = [];
        }

        const lRowNo = this._drawArcsUntilRow(pFrameNo);
        this._fillDownWithEmptyArcs(lRowNo);

        return this.AST;
    }

    /*
     * calculates the number of "frames" in the current AST
     * --> does not yet cater for recursive structures
     */
    private _calculateLength(pThing: any): number {
        let lRetval = 1; /* separate frame for entities */
        if (pThing.arcs) {
            lRetval = pThing.arcs.reduce((pSum, pArcRow) => /*
                * inner itself counts for two arcs (one extra for
                * drawing the bottom), but for one frame)
                */
            pSum + ((Boolean(pArcRow[0].arcs) ? this._calculateLength(pArcRow[0]) : pArcRow.length)), lRetval);
        }
        return lRetval;
    }

    /*
     * returns the number of rows for a given AST (/ AST snippet)
     */
    private _calcNumberOfRows(pThing) {
        let lRetval = 0;
        if (pThing.arcs) {
            lRetval = pThing.arcs.reduce(
                (pSum, pArcRow) => pSum + (Boolean(pArcRow[0].arcs) ? this._calcNumberOfRows(pArcRow[0]) + 2 : 1),
                lRetval,
            );
        }
        return lRetval;
    }
}

export default {
    FrameFactory,
};
