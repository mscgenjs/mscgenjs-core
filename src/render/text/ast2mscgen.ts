import aggregatekind from "../astmassage/aggregatekind";
import {XuAdaptor} from "./ast2xu";

export class MscGenAdaptor extends XuAdaptor {

    public init(pConfig = {}) {
        super.init(
            Object.assign(
                {
                    supportedOptions : ["hscale", "width", "arcgradient", "wordwraparcs"],
                    supportedEntityAttributes : [
                        "label", "idurl", "id", "url",
                        "linecolor", "textcolor", "textbgcolor",
                        "arclinecolor", "arctextcolor", "arctextbgcolor", "arcskip",
                    ],
                    supportedArcAttributes : [
                        "label", "idurl", "id", "url",
                        "linecolor", "textcolor", "textbgcolor",
                        "arclinecolor", "arctextcolor", "arctextbgcolor", "arcskip",
                    ],
                    inline : {
                        opener : `;${this.eol}`,
                        closer : "#",
                    },
                },
                pConfig,
            ),
        );
    }

    protected renderKind(pKind) {
        if ("inline_expression" === aggregatekind(pKind)) {
            return "--";
        }
        return pKind;
    }

    protected optionIsValid(pOption) {
        if (Boolean(pOption.value) && typeof (pOption.value) === "string") {
            return pOption.value.toLowerCase() !== "auto";
        }
        return true;
    }
}

export default {
    render: (pAST, pMinimal) => {
        const lAdaptor = new MscGenAdaptor(pMinimal);
        return lAdaptor.render(pAST);
    },
};
/*
 This file is part of mscgen_js.

 mscgen_js is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 mscgen_js is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with mscgen_js.  If not, see <http://www.gnu.org/licenses/>.
 */
