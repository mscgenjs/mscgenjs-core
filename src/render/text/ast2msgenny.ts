import type { ISequenceChart } from "../../parse/mscgenjsast";
import { XuAdaptor } from "./ast2xu";

class MsGennyAdaptor extends XuAdaptor {
  public init() {
    super.init({
      supportedEntityAttributes: ["label"],
      supportedArcAttributes: ["label"],
      program: {
        opener: "",
        closer: "",
      },
      option: {
        opener: "",
        separator: `,${this.eol}`,
        closer: `;${this.eol}${this.eol}`,
      },
      entity: {
        opener: "",
        separator: `,${this.eol}`,
        closer: `;${this.eol}${this.eol}`,
      },
      arcline: {
        opener: "",
        separator: `,${this.eol}`,
        closer: `;${this.eol}`,
      },
      inline: {
        opener: ` {${this.eol}`,
        closer: "}",
      },
      attribute: {
        opener: "",
        separator: "",
        closer: "",
      },
    });
  }

  protected renderEntityName(pString: string) {
    return this.entityNameIsQuotable(pString) ? `"${pString}"` : pString;
  }

  protected renderAttribute(pAttribute) {
    let lRetVal = "";
    if (pAttribute.name && pAttribute.value) {
      lRetVal += ` : "${pAttribute.value}"`;
    }
    return lRetVal;
  }

  private entityNameIsQuotable(pString: string) {
    const lMatchResult = pString.match(/[^;, "\t\n\r=\-><:{*]+/gi);
    if (lMatchResult) {
      return lMatchResult.length !== 1;
    } else {
      return pString !== "*";
    }
  }
}

export const render = (pAST: ISequenceChart) => {
  const lAdaptor = new MsGennyAdaptor();
  return lAdaptor.render(pAST);
};
/*
 This file is part of mscgen_js.

 mscgen_js is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 mscgen_js is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with mscgen_js.  If not, see <http://www.gnu.org/licenses/>.
 */
