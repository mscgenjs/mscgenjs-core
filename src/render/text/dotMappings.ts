/**
 * Defines several mappings of arckinds to agregations
 *
 */
const KIND2ARROW = Object.freeze({
    "->" : "rvee",
    "<->" : "rvee",
    "=>" : "normal",
    "<=>" : "normal",
    "-x" : "oinvonormal",
});
const KIND2SHAPE = Object.freeze({
    box : "box",
    abox : "hexagon",
    rbox : "box",
    note : "note",
});
const KIND2STYLE = Object.freeze({
    ">>" : "dashed",
    "<<>>" : "dashed",
    ".." : "dashed",
    ":>" : "bold",
    "<:>" : "bold",
    "::" : "bold",
    "rbox" : "rounded",
});

export default {
    getArrow(pKey) { return KIND2ARROW[pKey]; },
    getShape(pKey) { return KIND2SHAPE[pKey]; },
    getStyle(pKey) { return KIND2STYLE[pKey]; },
};

/* The 'generic object injection sink' is to a frozen object,
   attempts to modify it will be moot => we can safely use the []
   notation
*/
/*
 This file is part of mscgen_js.

 mscgen_js is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 mscgen_js is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with mscgen_js.  If not, see <http://www.gnu.org/licenses/>.
 */
