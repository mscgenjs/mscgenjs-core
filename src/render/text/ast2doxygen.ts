import type { ISequenceChart } from "../../parse/mscgenjsast";
import { MscGenAdaptor } from "./ast2mscgen";

class DoxygenAdaptor extends MscGenAdaptor {
  public init() {
    const INDENT = "  ";
    const EOL = "\n";
    const LINE_PREFIX = " * ";

    super.init({
      program: {
        opener: `${LINE_PREFIX}\\msc${EOL}`,
        closer: `${LINE_PREFIX}\\endmsc`,
      },
      option: {
        opener: LINE_PREFIX + INDENT,
        separator: `,${EOL}${LINE_PREFIX}${INDENT}`,
        closer: `;${EOL}${LINE_PREFIX}${EOL}`,
      },
      entity: {
        opener: LINE_PREFIX + INDENT,
        separator: `,${EOL}${LINE_PREFIX}${INDENT}`,
        closer: `;${EOL}${LINE_PREFIX}${EOL}`,
      },
      arcline: {
        opener: LINE_PREFIX + INDENT,
        separator: `,${EOL}${LINE_PREFIX}${INDENT}`,
        closer: `;${EOL}`,
      },
      inline: {
        opener: `;${EOL}`,
        closer: `${LINE_PREFIX}#`,
      },
    });
  }

  public renderComments() {
    /* rendering comments within comments, that are eventually output
     * to doxygen html - don't think that's going to be necessary
     * or desired functionality. If it is remember to be able to
     * - have a solution for nested comments (otherwise: interesting results)
     * - have a solution for comments that have an other meaning (# this is
     *    a comment -> doxygen translates this as markdown title)
     * - handling languages different from c/ java/ d that have alternative
     *   comment/ documentation sections
     */
    return "";
  }
}

export const render = (pAST: ISequenceChart) => {
  const lAdaptor = new DoxygenAdaptor(false);
  return lAdaptor.render(pAST);
};
