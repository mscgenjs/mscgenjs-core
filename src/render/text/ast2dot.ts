import cloneDeep from "lodash/cloneDeep";
import {
  ArcKindAggregatedType,
  IArc,
  IEntity,
  ISequenceChart,
} from "../../parse/mscgenjsast";
import aggregatekind from "../astmassage/aggregatekind";
import asttransform from "../astmassage/asttransform";
import * as flatten from "../astmassage/flatten";
import wrap from "../textutensils/wrap";
import * as dotMappings from "./dotMappings";

const INDENT = "  ";
const MAX_TEXT_WIDTH = 40;
let gCounter = 0;

/* Attribute handling */
function renderString(pString: string): string {
  const lStringAry = wrap(pString.replace(/"/g, '\\"'), MAX_TEXT_WIDTH);
  let lString = lStringAry
    .slice(0, -1)
    .reduce((pPrev, pLine) => `${pPrev + pLine}\n`, "");
  lString += lStringAry.slice(-1);
  return lString;
}

function renderAttribute(pAttr, pString) {
  return `${pString}="${renderString(pAttr)}"`;
}

function pushAttribute(pArray, pAttr, pString) {
  if (Boolean(pAttr)) {
    pArray.push(renderAttribute(pAttr, pString));
  }
}

function translateAttributes(pThing: IEntity | IArc): string[] {
  return ["label", "color", "fontcolor", "fillcolor"]
    .filter((pSupportedAttr) => Boolean(pThing[pSupportedAttr]))
    .map((pSupportedAttr) =>
      renderAttribute(pThing[pSupportedAttr], pSupportedAttr)
    );
}

function renderAttributeBlock(pAttrs): string {
  let lRetVal = "";
  // no need to check whether there's > 0 attribute passed here:
  // - entities have a mandatory 'name' attribute,
  // - arcs have a mandatory 'kind' attribute
  lRetVal = pAttrs
    .slice(0, -1)
    .reduce((pPrev, pAttr) => `${pPrev + pAttr}, `, " [");
  lRetVal += `${pAttrs.slice(-1)}]`;
  return lRetVal;
}

/* Entity handling */
function renderEntityName(pString: string): string {
  return `"${pString}"`;
}

function renderEntity(pEntity: IEntity): string {
  return (
    renderEntityName(pEntity.name) +
    renderAttributeBlock(translateAttributes(pEntity))
  );
}

function renderEntities(pEntities: IEntity[]): string {
  return pEntities.reduce(
    (pPrev, pEntity) => `${pPrev + INDENT + renderEntity(pEntity)};\n`,
    ""
  );
}

/* ArcLine handling */
function counterizeLabel(pCounter: number, pLabel?: string): string {
  if (pLabel) {
    return `(${pCounter}) ${pLabel}`;
  } else {
    return `(${pCounter})`;
  }
}

function renderBoxArc(pArc: any, pCounter: number, pIndent: string): string {
  let lRetVal = "";
  const lBoxName = `box${pCounter.toString()}`;
  lRetVal += lBoxName;
  let lAttrs = translateAttributes(pArc);
  pushAttribute(lAttrs, dotMappings.getStyle(pArc.kind), "style");
  pushAttribute(lAttrs, dotMappings.getShape(pArc.kind), "shape");

  lRetVal += `${renderAttributeBlock(lAttrs)}\n${INDENT}${pIndent}`;

  lAttrs = [];
  pushAttribute(lAttrs, "dotted", "style");
  pushAttribute(lAttrs, "none", "dir");

  lRetVal += `${lBoxName} -- {${renderEntityName(pArc.from)},${renderEntityName(
    pArc.to
  )}}`;
  lRetVal += renderAttributeBlock(lAttrs);

  return lRetVal;
}

function renderRegularArc(
  pArc: any,
  pAggregatedKind: ArcKindAggregatedType,
  pCounter: number
): string {
  let lRetVal = "";
  pArc.label = counterizeLabel(pCounter, pArc.label);
  const lAttrs = translateAttributes(pArc);

  pushAttribute(lAttrs, dotMappings.getStyle(pArc.kind), "style");
  switch (pAggregatedKind) {
    case "directional":
      pushAttribute(lAttrs, dotMappings.getArrow(pArc.kind), "arrowhead");
      break;
    case "bidirectional":
      pushAttribute(lAttrs, dotMappings.getArrow(pArc.kind), "arrowhead");
      pushAttribute(lAttrs, dotMappings.getArrow(pArc.kind), "arrowtail");
      pushAttribute(lAttrs, "both", "dir");
      break;
    case "nondirectional":
      pushAttribute(lAttrs, "none", "dir");
      break;
    default:
      break;
  }

  if (!pArc.arcs) {
    lRetVal += `${renderEntityName(pArc.from)} `;
    lRetVal += "--";
    lRetVal += ` ${renderEntityName(pArc.to)}`;
    lRetVal += renderAttributeBlock(lAttrs);
  }
  return lRetVal;
}

function renderSingleArc(
  pArc: IArc,
  pCounter: number,
  pIndent: string
): string {
  let lRetVal = "";
  const lAggregatedKind = aggregatekind(pArc.kind);

  if (lAggregatedKind === "box") {
    lRetVal += renderBoxArc(pArc, pCounter, pIndent);
  } else {
    lRetVal += renderRegularArc(pArc, lAggregatedKind, pCounter);
  }
  return lRetVal;
}

function renderArc(pArc: IArc, pIndent: string): string {
  let lRetVal = "";

  if (pArc.from && pArc.to) {
    lRetVal += `${
      INDENT + pIndent + renderSingleArc(pArc, ++gCounter, pIndent)
    }\n`;
    if (pArc.arcs) {
      lRetVal += `${INDENT + pIndent}subgraph cluster_${gCounter.toString()}{`;
      // not checking for pArc.label because there's at least a counter in it
      // at this point
      lRetVal += `\n${INDENT}${pIndent} label="${pArc.kind}: ${pArc.label}" labeljust="l"\n`;
      lRetVal += renderArcLines(pArc.arcs, pIndent + INDENT);
      lRetVal += `${INDENT + pIndent}}\n`;
    }
  }
  return lRetVal;
}

function renderArcLines(pArcLines: IArc[][], pIndent: string): string {
  return pArcLines.reduce(
    (pPrevArcLine, pNextArcLine) =>
      pPrevArcLine +
      pNextArcLine.reduce(
        (pPrevArc, pNextArc) => pPrevArc + renderArc(pNextArc, pIndent),
        ""
      ),
    ""
  );
}

function explodeBroadcastArc(pEntities: IEntity[], pArc: IArc): IArc[] {
  return pEntities
    .filter((pEntity) => pArc.from !== pEntity.name)
    .map((pEntity) => {
      pArc.to = pEntity.name;
      return cloneDeep(pArc);
    });
}

/**
 * - Gives each entity a label
 * - Sets arc kinds from left to right where applicable
 * - pre-calculates colors from regular colors and arc*-colors
 */
function flattenMe(pAST: ISequenceChart): ISequenceChart {
  return explodeBroadcasts(
    asttransform(
      pAST,
      [flatten.nameAsLabel],
      [flatten.swapRTLArc, flatten.overrideColors]
    )
  );
}

/**
 * expands "broadcast" arcs to its individual counterparts
 * Example in mscgen:
 *     a -> *;
 * output:
 *     a -> b, a -> c, a -> d;
 */
export function explodeBroadcasts(pAST): ISequenceChart {
  if (pAST.arcs) {
    pAST.arcs.forEach((pArcRow, pArcRowIndex) => {
      pArcRow
        // this assumes swap has been done already
        // and "*" is in no 'from'  anymore
        .filter((pArc) => pArc.to === "*")
        .forEach((pArc: IArc, pArcIndex: number) => {
          /* save a clone of the broadcast arc attributes
           * and remove the original bc arc
           */
          const lOriginalBroadcastArc = cloneDeep(pArc);

          delete pAST.arcs[pArcRowIndex][pArcIndex];
          const lExplodedArcsAry = explodeBroadcastArc(
            pAST.entities,
            lOriginalBroadcastArc
          );

          pArcRow[pArcIndex] = lExplodedArcsAry.shift();
          pAST.arcs[pArcRowIndex] = pArcRow.concat(lExplodedArcsAry);
        });
    });
  }
  return pAST;
}

export function render(pAST: ISequenceChart): string {
  const lAST: ISequenceChart = flattenMe(cloneDeep(pAST));
  let lRetVal =
    "/* Sequence chart represented as a directed graph\n" +
    " * in the graphviz dot language (http://graphviz.org/)\n" +
    " *\n" +
    " * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)\n" +
    " */\n" +
    "\n" +
    "graph {\n";
  lRetVal += `${INDENT}rankdir=LR\n`;
  lRetVal += `${INDENT}splines=true\n`;
  lRetVal += `${INDENT}ordering=out\n`;
  lRetVal += `${INDENT}fontname="Helvetica"\n`;
  lRetVal += `${INDENT}fontsize="9"\n`;
  lRetVal += `${INDENT}node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]\n`;
  lRetVal += `${INDENT}edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]\n`;
  lRetVal += "\n";

  lRetVal += `${renderEntities(lAST.entities)}\n`;
  if (lAST.arcs) {
    gCounter = 0;
    lRetVal += renderArcLines(lAST.arcs, "");
  }
  return (lRetVal += "}");
}

/*
 This file is part of mscgen_js.

 mscgen_js is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 mscgen_js is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with mscgen_js.  If not, see <http://www.gnu.org/licenses/>.
 */
