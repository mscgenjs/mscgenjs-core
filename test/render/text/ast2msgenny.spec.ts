import type { ISequenceChart } from "../../../src/parse/mscgenjsast";
import * as renderer from "../../../src/render/text/ast2msgenny";
const fix = require("../../astfixtures.json");

describe("render/text/ast2msgenny", () => {
  describe("#renderAST() - mscgen classic compatible - simple syntax trees", () => {
    test("should, given a simple syntax tree, render a msgenny script", () => {
      expect(renderer.render(fix.astSimple)).toMatchSnapshot();
    });

    test("should wrap labels with a , in quotes", () => {
      const lAST = {
        entities: [
          {
            name: "a",
            label: "comma,",
          },
        ],
      } as ISequenceChart;
      expect(renderer.render(lAST)).toMatchSnapshot();
    });

    test("should wrap labels with a ; in quotes", () => {
      const lAST = {
        entities: [
          {
            name: "a",
            label: "semi; colon",
          },
        ],
      } as ISequenceChart;
      expect(renderer.render(lAST)).toMatchSnapshot();
    });

    test("should wrap entity names with a space in quotes", () => {
      const lAST = {
        entities: [
          {
            name: "space space",
          },
        ],
      } as ISequenceChart;
      expect(renderer.render(lAST)).toMatchSnapshot();
    });

    test("should not wrap the '*' pseudo entity", () => {
      const lAST = {
        entities: [
          {
            name: "a",
          },
          {
            name: "b",
          },
          {
            name: "c",
          },
        ],
        arcs: [
          [
            {
              kind: "=>>",
              from: "b",
              to: "*",
              label: "",
            },
          ],
        ],
      } as ISequenceChart;
      expect(renderer.render(lAST)).toMatchSnapshot();
    });

    test("should render options when they're in the syntax tree", () => {
      expect(renderer.render(fix.astOptions)).toMatchSnapshot();
    });

    test("should ignore all attributes, except label and name", () => {
      expect(renderer.render(fix.astAllAttributes)).toMatchSnapshot();
    });

    test("should preserve the comments at the start of the ast", () => {
      expect(renderer.render(fix.astWithPreComment)).toMatchSnapshot();
    });

    test("should correctly render parallel calls", () => {
      expect(renderer.render(fix.astSimpleParallel)).toMatchSnapshot();
    });
  });

  describe("#renderAST() - xu compatible", () => {
    test("alt only - render correct script", () => {
      expect(renderer.render(fix.astOneAlt)).toMatchSnapshot();
    });
    test("alt within loop - render correct script", () => {
      expect(renderer.render(fix.astAltWithinLoop)).toMatchSnapshot();
    });
    test("should correctly render empty inline expressions", () => {
      const lFixture = {
        meta: {
          extendedOptions: false,
          extendedArcTypes: true,
          extendedFeatures: true,
        },
        entities: [
          {
            name: "a",
          },
          {
            name: "b",
          },
        ],
        arcs: [
          [
            {
              kind: "opt",
              from: "a",
              to: "b",
              arcs: null,
            },
          ],
        ],
      } as unknown;
      expect(renderer.render(<ISequenceChart>lFixture)).toMatchSnapshot();
    });
    test("Does not put entities with mscgen keyword for a name in quotes", () => {
      expect(
        renderer.render(fix.entityWithMscGenKeywordAsName)
      ).toMatchSnapshot();
    });
  });
});
