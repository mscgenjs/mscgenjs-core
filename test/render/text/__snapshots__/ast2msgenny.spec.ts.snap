// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should correctly render parallel calls 1`] = `
"a,
b,
c;

b -> a : "{paral",
b =>> c : "lel}";
"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should ignore all attributes, except label and name 1`] = `
"a : "Label for A";

a <<=>> a : "Label for a <<=>> a";
"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should not wrap the '*' pseudo entity 1`] = `
"a,
b,
c;

b =>> *;
"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should preserve the comments at the start of the ast 1`] = `
"# pre comment
/* pre
 * multiline
 * comment
 */
a,
b;

a -> b;
"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should render options when they're in the syntax tree 1`] = `
"hscale="1.2",
width="800",
arcgradient="17",
wordwraparcs=true,
watermark="not in mscgen, available in xÃ¹ and msgenny";

a;

"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should wrap entity names with a space in quotes 1`] = `
""space space";

"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should wrap labels with a , in quotes 1`] = `
"a : "comma,";

"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should wrap labels with a ; in quotes 1`] = `
"a : "semi; colon";

"
`;

exports[`render/text/ast2msgenny #renderAST() - mscgen classic compatible - simple syntax trees should, given a simple syntax tree, render a msgenny script 1`] = `
"a,
"b space";

a => "b space" : "a simple script";
"
`;

exports[`render/text/ast2msgenny #renderAST() - xu compatible Does not put entities with mscgen keyword for a name in quotes 1`] = `
"note;

"
`;

exports[`render/text/ast2msgenny #renderAST() - xu compatible alt only - render correct script 1`] = `
"a,
b,
c;

a => b;
b alt c {
  b => c;
  c >> b;
};
"
`;

exports[`render/text/ast2msgenny #renderAST() - xu compatible alt within loop - render correct script 1`] = `
"a,
b,
c;

a => b;
a loop c : "label for loop" {
  b alt c : "label for alt" {
    b -> c : "-> within alt";
    c >> b : ">> within alt";
  };
  b >> a : ">> within loop";
};
a =>> a : "happy-the-peppy - outside";
...;
"
`;

exports[`render/text/ast2msgenny #renderAST() - xu compatible should correctly render empty inline expressions 1`] = `
"a,
b;

a opt b {
};
"
`;
